[#50235] Granularly update Share Modal's components on CRUD actions

aaron-contreras opened the pull request:
## Description
Instead of re-rendering the whole modal on every CRUD action, we can perform more granular turbo stream responses:
### Create
1. New shares are prepended to the list, while preserving the alphabetical order of all initially rendered shares.
2. Updating a share via the invite form, removes the original share on the list and prepends the updated share to the the list.
![Kapture 2023-10-05 at 10 37 34](https://github.com/opf/openproject/assets/61627014/1f78bf49-7e0b-475d-81f1-4d753030e98a)

### Update
1. Ability to respond with a `head :ok`, meaning no re-rendering is needed.
2. Tab indexes are preserved (big on accessibility)
![Kapture 2023-10-05 at 10 44 47](https://github.com/opf/openproject/assets/61627014/017a66d5-f223-48d4-9870-5d4af3b4efb8)

### Destroy
1. Performs a `:remove` turbo stream action on the removed share
2. Preserves tab index (big on accessibility again)
![Kapture 2023-10-05 at 10 47 14](https://github.com/opf/openproject/assets/61627014/8dbdb116-c30e-41f4-9ee8-da383b44c3f1)

## Notes
### Noteworthy commits
https://github.com/opf/openproject/pull/13845/commits/ecc7d4d360475f9cfb63c7ecfe85572ad94bab8e: When performing a turbo stream response for a View Component slot, we need a way to specify the `with_*` call or whatever slot invocation mechanism is used as the `component_wrapper` and not only be able to provide a tag name as it stood before this PR (e.g. plain `div` with the wrapper key as its ID and some attributes).

This commit extends `component_wrapper` which can optionally take a method name as a symbol, which should render the custom component_wrapper you would like (e.g. a view component slot call to `box.with_row`) and adds the `wrapper_key` as its `id`.
### Work Package
https://community.openproject.org/work_packages/50235

### TODO
When https://github.com/opf/openproject/pull/13843 is merged this should be ready to review, as this commit is based off that PR and not based on dev

aaron-contreras committed:
Style: Use `#none?` instead of `#empty?`

Reads a bit better.

aaron-contreras committed:
Extract `WorkPackages::Share::ShareRowComponent`

aaron-contreras committed:
Extract `WorkPackages::Share::ShareCounterComponent`

aaron-contreras committed:
Allow injecting `shared_users` into ModalBodyComponent

aaron-contreras committed:
Update `destroy` action to granularly send turbo stream responses

aaron-contreras committed:
Perform no DOM manipulation on successful updates via inline row form

We don't really need to perform any DOM manipulation when updating
via the inline row select dropdown. A `head :no_content` suffices.

aaron-contreras committed:
Prepend newest created row to the top of shares list

aaron-contreras committed:
Allow custom-rendered `component_wrapper`

We should be able to render a custom implementation of component_wrapper
in scenarios where we're sending turbo stream responses for a "Slot"
view component and wouldn't want to have a different DOM structure when
sending a "replace" action with an extra tag as compared to initial
renders.

aaron-contreras committed:
Alphabetize order of shares in modal

aaron-contreras committed:
Add specs to cover granular update sorting behavior

ulferts commented on app/components/work_packages/share/modal_body_component.rb:
```diff
@@ -45,20 +47,39 @@ def self.wrapper_key
+      end
+
+      # There is currently no available system argument for setting an id on the
+      # rendered <ul> tag that houses the row slots on Primer::Beta::BorderBox components.
```
It would help in the future to document the motivation for setting the id.

ulferts commented on app/controllers/work_packages/shares_controller.rb:
```diff
@@ -39,41 +39,77 @@ def index
+    )
+
+    update_via_turbo_stream(
+      component: WorkPackages::Share::ShareCounterComponent.new(count: @shared_users.size)
```
This only works because the `@shared_users` variable is set as a side effect in a method called by the same caller before this here is called. 

Using `find_shared_users` would be less surprising. Same applies to [further up](https://github.com/opf/openproject/pull/13845/files#diff-1044ba07b8889f8ff65fb7b1338806f8909221c74d285aadcb659334a3ca169dR95)

ulferts commented on app/controllers/work_packages/shares_controller.rb:
```diff
@@ -92,4 +128,12 @@ def find_role_ids(builtin_value)
+                      .having_entity_membership(@work_package)
+                      .includes(work_package_shares: :roles)
+                      .where(work_package_shares: { entity: @work_package })
+                      .ordered_by_name
```
For the use cases in here, where only the existence of any share is checked for, something like

```
Member.of_work_package(@work_package)
```
would suffice.

Given the uses cases, it might be even better to just memoize the count in this method.

aaron-contreras replied:
I like the idea of memoizing the count. It'd read better

ulferts commented on spec/features/work_packages/share_spec.rb:
```diff
@@ -134,15 +129,15 @@
+      share_modal.expect_shared_with(not_shared_yet_with_user, 'Comment', position: 2)
+      share_modal.expect_shared_with(shared_project_user, 'Edit', position: 3)
+      share_modal.expect_shared_with(current_user, position: 4, editable: false)
+      share_modal.expect_shared_with(view_user, 'View', position: 5)
```
👍 for those changes to the specs. Quite nice to have that behaviour safeguarded.

ulferts commented on app/components/work_packages/share/modal_body_component.rb:
```diff
@@ -45,20 +47,39 @@ def self.wrapper_key
+                          .having_entity_membership(@work_package)
+                          .includes(work_package_shares: :roles)
+                          .where(work_package_shares: { entity: @work_package })
+                          .ordered_by_name
```
This is a bit weird to me. The method is [called once like this](https://github.com/opf/openproject/pull/13845/files#diff-dcc2c760c3da0ae92ef0691162adad19cc81662a330b2a062405da3c68fa139fR41)

```
@shared_users = find_shared_users
```

I'd thus remove the variable assignment here.

ulferts reviewed (changes requested):
I like that change a lot. The behaviour is much nicer. 

I still have to get used to the way the controllers are written, somewhat stateful, but that is on me. 

The only tiny thing that needs handling is resetting the autocompleter after adding an item:

<img width="696" alt="image" src="https://github.com/opf/openproject/assets/617519/2bd567e9-4409-45f0-9a7a-f45240c086ac">

My statement somewhat contradicts what I wanted earlier, saying that I don't want to rerender the autocompleter. But to be precise, I only want the autocompleter to be cleared after the invite button is pressed, not rerendered/refocused. Although, resetting the focus to the autocompleter after pressing the invite would still be nice. And since the clear behaviour will need some js anyway, you might be able to throw that in as well :). But to be clear, the refocusing should not happen when removing/editing existing shares.

aaron-contreras committed:
Clarify reason for setting custom id on BorderBox target

aaron-contreras committed:
Memoize current member count on shares controller

aaron-contreras committed:
Remove unneccessary duplicate instance variable assignment

aaron-contreras committed:
Replace `InviteUserFormComponent` on creation/update via form

aaron-contreras commented:
@ulferts Thanks for the feedback 👍 I've addressed all your remarks.

As discussed in our chat today, we compromised on replacing the `InviteUserFormComponent` on create/updates via the form. At least until our direction with the auto-completers is clear.

ulferts reviewed (approved):
I like it. Feels good to be using it.

Well done @aaron-contreras
